clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -1481,6 +1481,15 @@ static void handleOwnershipAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
     break;
   }
 
+  // Allow only pointers to be return type for functions with ownership_returns
+  // attribute. This matches with current OwnershipAttr::Takes semantics
+  if (K == OwnershipAttr::Returns) {
+    if (!getFunctionOrMethodResultType(D)->isPointerType()) {
+      S.Diag(AL.getLoc(), diag::err_ownership_takes_return_type) << AL;
+      return;
+    }
+  }
```

Comment:
```suggestion
  if (K == OwnershipAttr::Returns && 
     !getFunctionOrMethodResultType(D)->isPointerType()) {
    S.Diag(AL.getLoc(), diag::err_ownership_takes_return_type) << AL;
    return;
  }
```
(combining the two predicates, may need reformatting.)